File: backend/src/routes/universities.ts
/** Public university search and detail routes. */

import { Router } from 'express'
import {
  findSubjectCombination,
  programMatchesCombination,
  specialtiesForCombination,
  SUBJECT_COMBINATIONS,
} from '../data/subject-combinations.js'
import {
  findUniversityById,
  getProgramsByUniversity,
  listCities,
  listSpecialties,
  listUniversitiesFiltered,
  withPrograms,
} from '../db.js'
import type { ProgramRow } from '../types.js'

export const universitiesRouter = Router()

/** Filter programs by ENT score, subject combination, and specialty name. */
function filterPrograms(
  programs: ProgramRow[],
  entScore: number,
  combinationId?: string,
  specialty?: string,
): ProgramRow[] {
  const specialtyLower = specialty?.trim().toLowerCase()

  return programs.filter((program) => {
    if (program.min_ent_score > entScore) return false
    if (combinationId && !programMatchesCombination(program.gop_code, combinationId)) return false
    if (specialtyLower && program.name.toLowerCase() !== specialtyLower) return false
    return true
  })
}

/** Build list item with program match counts for ENT search results. */
function toListItem(
  university: ReturnType<typeof withPrograms>,
  entScore: number,
  combinationId?: string,
  specialty?: string,
) {
  const matched = filterPrograms(university.programs, entScore, combinationId, specialty)
  return {
    id: university.id,
    name: university.name,
    city: university.city,
    institution_type: university.institution_type,
    min_ent_score: university.min_ent_score,
    website: university.website,
    program_count: university.programs.length,
    matching_programs: matched.length,
  }
}

/** Search universities/colleges available for the given ENT score and filters. */
universitiesRouter.get('/search', (req, res) => {
  const entScore = Number(req.query.ent_score)
  if (Number.isNaN(entScore) || entScore < 0 || entScore > 140) {
    res.status(422).json({ detail: 'ent_score must be between 0 and 140' })
    return
  }

  const city = typeof req.query.city === 'string' ? req.query.city : undefined
  const institutionType =
    typeof req.query.institution_type === 'string' ? req.query.institution_type : undefined
  const combination =
    typeof req.query.combination === 'string' ? req.query.combination : undefined
  const specialty = typeof req.query.specialty === 'string' ? req.query.specialty : undefined

  if (combination && !findSubjectCombination(combination)) {
    res.status(422).json({ detail: 'Unknown subject combination' })
    return
  }

  const rows = listUniversitiesFiltered(city, institutionType)
  const matches = rows
    .map((row) => withPrograms(row))
    .filter((uni) => filterPrograms(uni.programs, entScore, combination, specialty).length > 0)
    .map((uni) => toListItem(uni, entScore, combination, specialty))

  res.json({
    ent_score: entScore,
    combination: combination ?? null,
    specialty: specialty ?? null,
    total_matches: matches.length,
    universities: matches,
  })
})

/** Return ENT profile subject combinations for the search form. */
universitiesRouter.get('/combinations', (_req, res) => {
  res.json(SUBJECT_COMBINATIONS.map(({ id, label }) => ({ id, label })))
})

/** Return specialty names for the search form, optionally scoped to a combination. */
universitiesRouter.get('/specialties', (req, res) => {
  const combination =
    typeof req.query.combination === 'string' ? req.query.combination : undefined

  if (combination) {
    if (!findSubjectCombination(combination)) {
      res.status(422).json({ detail: 'Unknown subject combination' })
      return
    }
    res.json(specialtiesForCombination(combination))
    return
  }

  res.json(listSpecialties())
})

/** Return distinct cities for frontend filters. */
universitiesRouter.get('/cities', (_req, res) => {
  res.json(listCities())
})

/** List all programs for a university without ENT filtering. */
universitiesRouter.get('/:universityId/programs', (req, res) => {
  const universityId = Number(req.params.universityId)
  const university = findUniversityById(universityId)
  if (!university) {
    res.status(404).json({ detail: 'University not found' })
    return
  }
  res.json(getProgramsByUniversity(universityId))
})

/** Return one university; optionally filter programs by ENT score and filters. */
universitiesRouter.get('/:universityId', (req, res) => {
  const universityId = Number(req.params.universityId)
  const entScoreRaw = req.query.ent_score
  const entScore = entScoreRaw != null ? Number(entScoreRaw) : undefined
  const combination =
    typeof req.query.combination === 'string' ? req.query.combination : undefined
  const specialty = typeof req.query.specialty === 'string' ? req.query.specialty : undefined

  const university = findUniversityById(universityId)
  if (!university) {
    res.status(404).json({ detail: 'University not found' })
    return
  }

  const full = withPrograms(university)
  if (entScore != null && !Number.isNaN(entScore)) {
    full.programs = filterPrograms(full.programs, entScore, combination, specialty)
  }

  res.json(full)
})

File: backend/app/routers/universities.py
"""Public university search by ENT score."""

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func
from sqlalchemy.orm import Session, joinedload

from app.database import get_db
from app.models import Program, University
from app.schemas import EntSearchResponse, ProgramRead, UniversityListItem, UniversityRead

router = APIRouter(prefix="/api/universities", tags=["universities"])


def _build_list_item(university: University, ent_score: int | None) -> UniversityListItem:
    """Map a university ORM row to a list item with program match counts."""
    program_count = len(university.programs)
    matching_programs = 0
    if ent_score is not None:
        matching_programs = sum(1 for program in university.programs if program.min_ent_score <= ent_score)
    return UniversityListItem(
        id=university.id,
        name=university.name,
        city=university.city,
        institution_type=university.institution_type,
        min_ent_score=university.min_ent_score,
        website=university.website,
        program_count=program_count,
        matching_programs=matching_programs,
    )


@router.get("/search", response_model=EntSearchResponse)
def search_by_ent(
    ent_score: int = Query(..., ge=0, le=140, description="Балл ЕНТ абитуриента"),
    city: str | None = Query(default=None, description="Фильтр по городу"),
    institution_type: str | None = Query(default=None, description="university или college"),
    db: Session = Depends(get_db),
):
    """Return universities and colleges where the user ENT meets minimum thresholds."""
    query = db.query(University).options(joinedload(University.programs))
    if city:
        query = query.filter(func.lower(University.city) == city.lower())
    if institution_type:
        query = query.filter(University.institution_type == institution_type)

    universities = query.order_by(University.min_ent_score.desc()).all()
    matches: list[UniversityListItem] = []
    for university in universities:
        has_matching_program = any(program.min_ent_score <= ent_score for program in university.programs)
        university_ok = university.min_ent_score <= ent_score
        if university_ok or has_matching_program:
            matches.append(_build_list_item(university, ent_score))

    return EntSearchResponse(ent_score=ent_score, total_matches=len(matches), universities=matches)


@router.get("/cities", response_model=list[str])
def list_cities(db: Session = Depends(get_db)):
    """Return distinct cities for filter dropdowns."""
    rows = db.query(University.city).distinct().order_by(University.city).all()
    return [row[0] for row in rows]


@router.get("/{university_id}", response_model=UniversityRead)
def get_university(
    university_id: int,
    ent_score: int | None = Query(default=None, ge=0, le=140),
    db: Session = Depends(get_db),
):
    """Return one university with programs; optionally highlight ENT eligibility."""
    university = (
        db.query(University)
        .options(joinedload(University.programs))
        .filter(University.id == university_id)
        .first()
    )
    if not university:
        raise HTTPException(status_code=404, detail="University not found")

    data = UniversityRead.model_validate(university)
    if ent_score is not None:
        data.programs = [program for program in data.programs if program.min_ent_score <= ent_score]
    return data


@router.get("/{university_id}/programs", response_model=list[ProgramRead])
def list_programs(university_id: int, db: Session = Depends(get_db)):
    """List all programs for a university."""
    programs = db.query(Program).filter(Program.university_id == university_id).order_by(Program.min_ent_score).all()
    return programs
